libbeat/common/kubernetes/metadata/metadata.go

```diff
@@ -31,11 +48,32 @@ type MetaGen interface {
 	Generate(kubernetes.Resource, ...FieldOptions) common.MapStr
 	// GenerateFromName generates metadata for a given resource based on it's name
 	GenerateFromName(string, ...FieldOptions) common.MapStr
+	// GenerateK8s generates metadata for a given resource
+	GenerateK8s(kubernetes.Resource, ...FieldOptions) common.MapStr
+	// GenerateK8s generates metadata for a given resource
+	GenerateECS(kubernetes.Resource) common.MapStr
```

Comment:
This interface can be confusing. When implementing a new generator, how to decide what to generate with each one of these four Generate methods?

Maybe this is a reason in favour of trying to leverage `FieldOptions`, and do the fetch of the cluster metadata in a different object.

Reply:
Sorry for not updating the docstrings earlier, this makes it even more confusing.
I can understand the concern here, I had similar concerns while trying to put all these together. However the key part here is that we need something new while we are adding more ECS fields as part of metadata. Today `Generate` methods return a map that consists of metadata that are supposed to be stored under `kubernetes.*` prefix, which prefix is added in top level where `Generate` method is called. What happen though when we want to add metadata through these generators that are supposed to be stored in ECS fields and hence not under `kubernetes.*` key? This happened already more or less with `container.*` fields. The same will happen with `kubernetes.namespace` too when we will decide to move it to `orchestrator.namespace`. So what I'm thinking of is that we might need a higher level `Generate` method which will be responsible to fetch all k8s related metadata in a final format. This is why I introduced `GenerateK8s` (which is former `Generate`) and `GenerateECS` so as to combine these to two in `Generate` and return a complete set of metadata consists not only of `kubernetes.*` fields but also ECS fields. Also keep in mind that we already have a fairly complex mechanism there to handle sub-resource's metadata, and `Generate` can be called directly for Pod's metadata for example or called extra for Node's metadata under Pod's metadata. I'm open to discuss it more (here or offline if you prefer) so as to conclude to a more future proof implementation for this :).

Reply:
Umm, ok to the separation in two methods for the different sets of fields, and to make `Generate` to return root-level fields. 
But `GenerateK8s` and `GenerateECS` are currently called only by `Generate`, do they need to be in the interface?

Reply:
I think that we still need those methods to be exposed because of the need to directly call them not from top level but from inside `MetaGenerator`s. For example `PodMetaGenerator` needs to call `GenerateK8s` of generic `Resource` directly so as to get k8s specific metadata at https://github.com/elastic/beats/blob/3b2761f1b8e8634dc2cad4353563f31e5aec4cf3/libbeat/common/kubernetes/metadata/pod.go#L84.